File: Ruby/Easy/13. Roman to Integer.rb
# @param {String} s
# @return {Integer}
def roman_to_int(s)
    hash={'I'=>1,'V'=>5,'X'=>10,'L'=>50,'C'=>100,'D'=>500,'M'=>1000}
    s=s.reverse()
    ans=hash[s[0]]
    last=ans
    for idx in 1...s.length()do
        a=hash[s[idx]]
        if last>a
            ans-=a
        else
            ans+=a
        end
        last=a
    end
    return ans
end

File: Python/13. Roman to Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        value_map = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
        ans = 0
        l = len(s)

        for idx in range(l - 1, -1, -1):
            sign = 1
            if idx < l - 1:
                sign = 1 if value_map[s[idx]] >= value_map[s[idx + 1]] else -1
            ans += sign * value_map[s[idx]]

        return ans
